File: tests/wasm/test_cycle_channel_oscillator.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadTestData, assertArrayClose } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualRma(values, length) {
    const out = new Float64Array(values.length);
    out.fill(NaN);
    let total = 0;
    let count = 0;
    let current = NaN;
    for (let i = 0; i < values.length; i++) {
        const value = values[i];
        if (!Number.isFinite(value)) {
            continue;
        }
        if (count < length) {
            total += value;
            count += 1;
            if (count === length) {
                current = total / length;
                out[i] = current;
            }
        } else {
            current = current + (value - current) / length;
            count += 1;
            out[i] = current;
        }
    }
    return out;
}

function manualAtr(high, low, close, length) {
    const tr = new Float64Array(close.length);
    tr.fill(NaN);
    let prevClose = null;
    for (let i = 0; i < close.length; i++) {
        if (!(Number.isFinite(high[i]) && Number.isFinite(low[i]) && Number.isFinite(close[i]))) {
            continue;
        }
        if (prevClose === null) {
            tr[i] = high[i] - low[i];
        } else {
            tr[i] = Math.max(
                high[i] - low[i],
                Math.abs(high[i] - prevClose),
                Math.abs(low[i] - prevClose),
            );
        }
        prevClose = close[i];
    }
    return manualRma(tr, length);
}

function delayedOrCurrent(values, idx, delay, current) {
    if (idx >= delay && Number.isFinite(values[idx - delay])) {
        return values[idx - delay];
    }
    return current;
}

function manualCycleChannelOscillator(source, high, low, close, shortCycleLength, mediumCycleLength, mediumMultiplier) {
    const shortPeriod = Math.floor(shortCycleLength / 2);
    const mediumPeriod = Math.floor(mediumCycleLength / 2);
    const shortDelay = Math.floor(shortPeriod / 2);
    const mediumDelay = Math.floor(mediumPeriod / 2);
    const shortMa = manualRma(source, shortPeriod);
    const mediumMa = manualRma(source, mediumPeriod);
    const mediumAtr = manualAtr(high, low, close, mediumPeriod);
    const fast = new Float64Array(source.length);
    const slow = new Float64Array(source.length);
    fast.fill(NaN);
    slow.fill(NaN);

    for (let i = 0; i < source.length; i++) {
        if (!(Number.isFinite(source[i]) && Number.isFinite(high[i]) && Number.isFinite(low[i]) && Number.isFinite(close[i]))) {
            continue;
        }
        const mediumCenter = delayedOrCurrent(mediumMa, i, mediumDelay, source[i]);
        const shortCenter = delayedOrCurrent(shortMa, i, shortDelay, source[i]);
        const offset = mediumMultiplier * mediumAtr[i];
        const denom = 2 * offset;
        if (Number.isFinite(denom) && denom !== 0) {
            const mediumBottom = mediumCenter - offset;
            fast[i] = (source[i] - mediumBottom) / denom;
            slow[i] = (shortCenter - mediumBottom) / denom;
        }
    }

    return { fast, slow };
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('cycle_channel_oscillator_js matches manual reference', () => {
    const n = 160;
    const source = new Float64Array(testData.close.slice(0, n));
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const result = wasm.cycle_channel_oscillator_js(source, high, low, close, 10, 30, 1.0, 3.0);
    const expected = manualCycleChannelOscillator(source, high, low, close, 10, 30, 3.0);
    assertArrayClose(result.fast, expected.fast, 1e-12, 'fast mismatch');
    assertArrayClose(result.slow, expected.slow, 1e-12, 'slow mismatch');
});

test('cycle_channel_oscillator_batch_js first row matches single', () => {
    const n = 96;
    const source = new Float64Array(testData.close.slice(0, n));
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const batch = wasm.cycle_channel_oscillator_batch_js(source, high, low, close, {
        short_cycle_length_range: [10, 10, 0],
        medium_cycle_length_range: [30, 32, 2],
        short_multiplier_range: [1.0, 1.0, 0.0],
        medium_multiplier_range: [3.0, 3.0, 0.0],
    });
    const single = wasm.cycle_channel_oscillator_js(source, high, low, close, 10, 30, 1.0, 3.0);
    assert.strictEqual(batch.rows, 2);
    assert.strictEqual(batch.cols, n);
    assert.deepStrictEqual(batch.short_cycle_lengths, [10, 10]);
    assert.deepStrictEqual(batch.medium_cycle_lengths, [30, 32]);
    assertArrayClose(batch.fast.slice(0, n), single.fast, 1e-12, 'batch fast mismatch');
    assertArrayClose(batch.slow.slice(0, n), single.slow, 1e-12, 'batch slow mismatch');
});

test('cycle_channel_oscillator_into pointer path matches safe API', () => {
    const n = 128;
    const source = new Float64Array(testData.close.slice(0, n));
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const safe = wasm.cycle_channel_oscillator_js(source, high, low, close, 10, 30, 1.0, 3.0);

    const sourcePtr = wasm.allocate_f64_array(n);
    const highPtr = wasm.allocate_f64_array(n);
    const lowPtr = wasm.allocate_f64_array(n);
    const closePtr = wasm.allocate_f64_array(n);
    const outFastPtr = wasm.cycle_channel_oscillator_alloc(n);
    const outSlowPtr = wasm.cycle_channel_oscillator_alloc(n);

    try {
        wasm.write_f64_array(sourcePtr, source);
        wasm.write_f64_array(highPtr, high);
        wasm.write_f64_array(lowPtr, low);
        wasm.write_f64_array(closePtr, close);
        wasm.cycle_channel_oscillator_into(
            sourcePtr, highPtr, lowPtr, closePtr, outFastPtr, outSlowPtr, n, 10, 30, 1.0, 3.0,
        );
        const gotFast = wasm.read_f64_array(outFastPtr, n);
        const gotSlow = wasm.read_f64_array(outSlowPtr, n);
        assertArrayClose(gotFast, safe.fast, 1e-12, 'pointer fast mismatch');
        assertArrayClose(gotSlow, safe.slow, 1e-12, 'pointer slow mismatch');
    } finally {
        wasm.cycle_channel_oscillator_free(outFastPtr, n);
        wasm.cycle_channel_oscillator_free(outSlowPtr, n);
        wasm.deallocate_f64_array(sourcePtr);
        wasm.deallocate_f64_array(highPtr);
        wasm.deallocate_f64_array(lowPtr);
        wasm.deallocate_f64_array(closePtr);
    }
});

test('cycle_channel_oscillator_batch_into returns row count', () => {
    const n = 96;
    const rows = 2;
    const source = new Float64Array(testData.close.slice(0, n));
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));

    const sourcePtr = wasm.allocate_f64_array(n);
    const highPtr = wasm.allocate_f64_array(n);
    const lowPtr = wasm.allocate_f64_array(n);
    const closePtr = wasm.allocate_f64_array(n);
    const outFastPtr = wasm.cycle_channel_oscillator_alloc(rows * n);
    const outSlowPtr = wasm.cycle_channel_oscillator_alloc(rows * n);

    try {
        wasm.write_f64_array(sourcePtr, source);
        wasm.write_f64_array(highPtr, high);
        wasm.write_f64_array(lowPtr, low);
        wasm.write_f64_array(closePtr, close);
        const returnedRows = wasm.cycle_channel_oscillator_batch_into(
            sourcePtr,
            highPtr,
            lowPtr,
            closePtr,
            outFastPtr,
            outSlowPtr,
            n,
            10, 10, 0,
            30, 32, 2,
            1.0, 1.0, 0.0,
            3.0, 3.0, 0.0,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.cycle_channel_oscillator_free(outFastPtr, rows * n);
        wasm.cycle_channel_oscillator_free(outSlowPtr, rows * n);
        wasm.deallocate_f64_array(sourcePtr);
        wasm.deallocate_f64_array(highPtr);
        wasm.deallocate_f64_array(lowPtr);
        wasm.deallocate_f64_array(closePtr);
    }
});

test('cycle_channel_oscillator_js rejects invalid length', () => {
    const n = 64;
    const source = new Float64Array(testData.close.slice(0, n));
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    assert.throws(() => {
        wasm.cycle_channel_oscillator_js(source, high, low, close, 1, 30, 1.0, 3.0);
    }, /Invalid cycle length/);
});

File: tests/python/test_cycle_channel_oscillator.py
import numpy as np
import pytest

from test_utils import load_test_data

try:
    import my_project as mp
except ImportError:
    pytest.skip(
        "Python module not built. Run 'maturin develop --features python' first",
        allow_module_level=True,
    )


def manual_rma(values: np.ndarray, length: int) -> np.ndarray:
    out = np.full(values.shape[0], np.nan, dtype=np.float64)
    total = 0.0
    count = 0
    current = np.nan
    for i, value in enumerate(values):
        if not np.isfinite(value):
            continue
        if count < length:
            total += value
            count += 1
            if count == length:
                current = total / float(length)
                out[i] = current
        else:
            current = current + (value - current) / float(length)
            count += 1
            out[i] = current
    return out


def manual_atr(high: np.ndarray, low: np.ndarray, close: np.ndarray, length: int) -> np.ndarray:
    tr = np.full(close.shape[0], np.nan, dtype=np.float64)
    prev_close = None
    for i in range(close.shape[0]):
        if not (np.isfinite(high[i]) and np.isfinite(low[i]) and np.isfinite(close[i])):
            continue
        if prev_close is None:
            tr[i] = high[i] - low[i]
        else:
            tr[i] = max(
                high[i] - low[i],
                abs(high[i] - prev_close),
                abs(low[i] - prev_close),
            )
        prev_close = close[i]
    return manual_rma(tr, length)


def delayed_or_current(values: np.ndarray, idx: int, delay: int, current: float) -> float:
    if idx >= delay and np.isfinite(values[idx - delay]):
        return float(values[idx - delay])
    return current


def manual_cycle_channel_oscillator(
    source: np.ndarray,
    high: np.ndarray,
    low: np.ndarray,
    close: np.ndarray,
    short_cycle_length: int,
    medium_cycle_length: int,
    medium_multiplier: float,
):
    short_period = short_cycle_length // 2
    medium_period = medium_cycle_length // 2
    short_delay = short_period // 2
    medium_delay = medium_period // 2

    short_ma = manual_rma(source, short_period)
    medium_ma = manual_rma(source, medium_period)
    medium_atr = manual_atr(high, low, close, medium_period)

    fast = np.full(source.shape[0], np.nan, dtype=np.float64)
    slow = np.full(source.shape[0], np.nan, dtype=np.float64)
    for i in range(source.shape[0]):
        if not (
            np.isfinite(source[i])
            and np.isfinite(high[i])
            and np.isfinite(low[i])
            and np.isfinite(close[i])
        ):
            continue
        medium_center = delayed_or_current(medium_ma, i, medium_delay, float(source[i]))
        short_center = delayed_or_current(short_ma, i, short_delay, float(source[i]))
        offset = medium_multiplier * medium_atr[i]
        denom = 2.0 * offset
        if np.isfinite(denom) and denom != 0.0:
            medium_bottom = medium_center - offset
            fast[i] = (source[i] - medium_bottom) / denom
            slow[i] = (short_center - medium_bottom) / denom
    return fast, slow


class TestCycleChannelOscillator:
    def test_manual_reference_matches_binding(self):
        data = load_test_data()
        n = 160
        source = np.asarray(data["close"][:n], dtype=np.float64)
        high = np.asarray(data["high"][:n], dtype=np.float64)
        low = np.asarray(data["low"][:n], dtype=np.float64)
        close = np.asarray(data["close"][:n], dtype=np.float64)
        expected_fast, expected_slow = manual_cycle_channel_oscillator(
            source, high, low, close, 10, 30, 3.0
        )
        result = mp.cycle_channel_oscillator(
            source,
            high,
            low,
            close,
            short_cycle_length=10,
            medium_cycle_length=30,
            short_multiplier=1.0,
            medium_multiplier=3.0,
        )
        np.testing.assert_allclose(result["fast"], expected_fast, rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(result["slow"], expected_slow, rtol=0.0, atol=1e-12, equal_nan=True)

    def test_stream_matches_batch(self):
        data = load_test_data()
        n = 128
        source = np.asarray(data["close"][:n], dtype=np.float64)
        high = np.asarray(data["high"][:n], dtype=np.float64)
        low = np.asarray(data["low"][:n], dtype=np.float64)
        close = np.asarray(data["close"][:n], dtype=np.float64)
        batch = mp.cycle_channel_oscillator(source, high, low, close)
        stream = mp.CycleChannelOscillatorStream()
        streamed = np.array(
            [stream.update(float(s), float(h), float(l), float(c)) for s, h, l, c in zip(source, high, low, close)]
        )
        np.testing.assert_allclose(streamed[:, 0], batch["fast"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 1], batch["slow"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_batch_metadata_and_first_row_match_single(self):
        data = load_test_data()
        n = 96
        source = np.asarray(data["close"][:n], dtype=np.float64)
        high = np.asarray(data["high"][:n], dtype=np.float64)
        low = np.asarray(data["low"][:n], dtype=np.float64)
        close = np.asarray(data["close"][:n], dtype=np.float64)
        result = mp.cycle_channel_oscillator_batch(
            source,
            high,
            low,
            close,
            short_cycle_length_range=(10, 10, 0),
            medium_cycle_length_range=(30, 32, 2),
            short_multiplier_range=(1.0, 1.0, 0.0),
            medium_multiplier_range=(3.0, 3.0, 0.0),
        )
        assert result["rows"] == 2
        assert result["cols"] == n
        np.testing.assert_array_equal(result["short_cycle_lengths"], np.array([10, 10], dtype=np.uint64))
        np.testing.assert_array_equal(result["medium_cycle_lengths"], np.array([30, 32], dtype=np.uint64))

        single = mp.cycle_channel_oscillator(source, high, low, close)
        np.testing.assert_allclose(result["fast"][0], single["fast"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(result["slow"][0], single["slow"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_kernel_parity(self):
        data = load_test_data()
        n = 144
        source = np.asarray(data["close"][:n], dtype=np.float64)
        high = np.asarray(data["high"][:n], dtype=np.float64)
        low = np.asarray(data["low"][:n], dtype=np.float64)
        close = np.asarray(data["close"][:n], dtype=np.float64)
        auto = mp.cycle_channel_oscillator(source, high, low, close)
        scalar = mp.cycle_channel_oscillator(source, high, low, close, kernel="scalar")
        np.testing.assert_allclose(auto["fast"], scalar["fast"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(auto["slow"], scalar["slow"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_invalid_length(self):
        data = load_test_data()
        source = np.asarray(data["close"][:32], dtype=np.float64)
        high = np.asarray(data["high"][:32], dtype=np.float64)
        low = np.asarray(data["low"][:32], dtype=np.float64)
        close = np.asarray(data["close"][:32], dtype=np.float64)
        with pytest.raises(ValueError, match="Invalid cycle length"):
            mp.cycle_channel_oscillator(source, high, low, close, short_cycle_length=1)
